### ISSUE TITLE
Android connect fails with invalid protocol

### ISSUE THREAD

**Issue by hopchop** [anoved/mmserver#36]
MMSERVER works fine with latest IOS, but latest Android fails like this:

mmserver[6963]: [192.168.254.67] connected
mmserver[6963]: [192.168.254.67] disconnected (invalid protocol)
  0x0000:  43 4f 4e 4e 45 43 54 1e 6e 75 6c 6c 1e 41 6e 64    CONNECT.null.And
  0x0010:  72 6f 69 64 1e 41 6e 64 72 6f 69 64 1e 32 2e 30    roid.Android.2.0
  0x0020:  2e 36 1e 30 04                                     .6.0.

**Comment by anoved** [comment]
Thanks for the heads-up. I don't have an Android device, so I can't test this directly, but I can sort of see what's going wrong. Mmserver expects to receive a "CONNECT" message in a certain format when the client connects, and it looks like the Android client is using a different message format - see the "null.Android.Android.2.0.6.0" data appended to the CONNECT message.

**Comment by hopchop** [comment]
Hi Jim!

I have two Android devices with the latest version "Kitkat". They both have the same problem.

It seems like this matching in session.cpp fails:

if (!pcrecpp::RE("CONNECT\x1e(._?)\x1e(._?)\x1e(.*?)\x1e[0-9]\x1e[0-9]\x04").FullMatch(buffer,
                &password, &id, &name))
    {
        syslog(LOG_INFO, "[%s] disconnected (invalid protocol)", address.c_str());

```
    /* dump unhandled packets */
```

Maybe the message format is not different, just the initial handshake?

Kind regards,

Walter

Am 14.02.2014 um 19:12 schrieb Jim notifications@github.com:

> Thanks for the heads-up. I don't have an Android device, so I can't test this directly, but I can sort of see what's going wrong. Mmserver expects to receive a "CONNECT" message in a certain format when the client connects, and it looks like the Android client is using a different message format - see the "null.Android.Android.2.0" data appended to the CONNECT message.
> 
> —
> Reply to this email directly or view it on GitHub.

---

President
European Professional Beekeepers Association
Member of the Board of Directors
German Professional Beekeepers Association
Deutscher Berufs und Erwerbsimkerbund e.V
Mobile:  +49 151 58 565 444
Fax: +49 322 224 507 65
E-Mail: walter.haefeker@berufsimker.de
http://www.berufsimker.de

**Comment by anoved** [comment]
Yep. It appears the problem is mmserver expects an exactly two digit version number at the end of that handshake (the `[0-9]\x1e[0-9]`). It should probably not be so strict, as I'm guessing the Android client version number is something like `2.0.6.0`.

If you are able to make and test the required change to fix the match, I will gladly accept the patch. Otherwise I will try to do it when I get a chance (may be a day or two). Thanks!

**Comment by hopchop** [comment]
I'm not a regex wizard. If you suggest a few patterns I should try, I will modify the code on my machine, test it and let you know which one did the trick.

The app itself claims to be 2.0.6 in the Android app settings.

In oder to avoid running into this with every new release, maybe a setting in the config file would either allow the user to disable strict checking or to enter the version string from the app the server should expect.

Kind regards,

Walter

Am 14.02.2014 um 21:21 schrieb Jim notifications@github.com:

> Yep. It appears the problem is mmserver expects an exactly two digit version number at the end of that handshake (the [0-9]\x1e[0-9]). It should probably not be so strict, as I'm guessing the Android client version number is something like 2.0.6.0.
> 
> If you are able to make and test the required change to fix the match, I will gladly accept the patch. Otherwise I will try to do it when I get a chance (may be a day or two). Thanks!
> 
> —
> Reply to this email directly or view it on GitHub.

---

President
European Professional Beekeepers Association
Member of the Board of Directors
German Professional Beekeepers Association
Deutscher Berufs und Erwerbsimkerbund e.V
Mobile:  +49 151 58 565 444
Fax: +49 322 224 507 65
E-Mail: walter.haefeker@berufsimker.de
http://www.berufsimker.de

**Comment by anoved** [comment]
Mmserver doesn't do anything with the version number anyway, so I just changed it to disregard that part of the connection message. It still works with iOS and should hopefully fix the Android connection problem. Let me know if so!

**Comment by mrdrmmn** [comment]
I decided to give mmserver a try and ran into this problem (android client, server running on 64 bit ubuntu 13.10). I just rebuilt from master and while the "invalid protocol" message is gone, I now get a 'connected' message immediately followed by a 'disconnected (read failed: Resource temporarily unavailable)' message and then a 'session ended' message.

This may be completely unrelated (or not), but the avahi publishing is not working for me either (even though it is specified in the config).

**Comment by anoved** [comment]
Walter, can you confirm if the same thing happens for you as reported by therevmj?

For what it's worth, using the iOS client, I occasionally see the same "resource temporarily unavailable/session ended" error, followed immediately by a reconnect. Do I infer correctly that your client does not reconnect?

therevmj, can you connect by entering the server's IP and port number in the client app directly? I think that would help determine whether avahi problems are related to your connection error or are a separate issue.

**Comment by hopchop** [comment]
Hi JIm!

This is what I get as an endless loop until the app on Android crashes 
or I shut it down:

mmserver[31009]: [192.168.254.67] session ended
mmserver[31009]: [192.168.254.67] connected
mmserver[31009]: [192.168.254.67] device id: Android
mmserver[31009]: [192.168.254.67] device name: Android
mmserver[31009]: [192.168.254.67] disconnected (read failed: Die 
Ressource ist zur Zeit nicht verfügbar)
mmserver[31009]: [192.168.254.67] session ended
mmserver[31009]: [192.168.254.67] connected
mmserver[31009]: [192.168.254.67] device id: Android
mmserver[31009]: [192.168.254.67] device name: Android
mmserver[31009]: [192.168.254.67] disconnected (read failed: Die 
Ressource ist zur Zeit nicht verfügbar)
mmserver[31009]: [192.168.254.67] session ended
mmserver[31009]: [192.168.254.67] connected
mmserver[31009]: [192.168.254.67] device id: Android
mmserver[31009]: [192.168.254.67] device name: Android
mmserver[31009]: [192.168.254.67] disconnected (read failed: Die 
Ressource ist zur Zeit nicht verfügbar)
mmserver[31009]: [192.168.254.67] session ended

I'm running this on a thinkpad with German language setting.

Die Ressource ist zur Zeit nicht verfügbar means Resource currently 
unavailable.

Kind regards,

Walter

Am 18.02.2014 15:48, schrieb Jim:

**Comment by anoved** [comment]
Thanks for the confirmation.

I don't know why this happens, so I've added a few debug statements to print the last data received from the app when the "read failed" errors occur. I hope this will clarify what is happening. The Android app's behavior is evidently not identical to the iOS app, but it's difficult for me to determine how so.

**Comment by hopchop** [comment]
Hi Jim!

Here is the output with the debug code:

mmserver[25175]: [192.168.254.67] session ended
mmserver[25175]: [192.168.254.67] connected
mmserver[25175]: [192.168.254.67] device id: Android
mmserver[25175]: [192.168.254.67] device name: Android
mmserver[25175]: [192.168.254.67] disconnected (read failed: Die 
Ressource ist zur Zeit nicht verfügbar)
   0x0000:  43 4f 4e 4e 45 43 54 1e 1e 41 6e 64 72 6f 69 64 CONNECT..Android
   0x0010:  1e 41 6e 64 72 6f 69 64 1e 32 2e 30 2e 36 1e 30 .Android.2.0.6.0
   0x0020:  04 .
mmserver[25175]: [192.168.254.67] session ended
mmserver[25175]: [192.168.254.67] connected
mmserver[25175]: [192.168.254.67] device id: Android
mmserver[25175]: [192.168.254.67] device name: Android
mmserver[25175]: [192.168.254.67] disconnected (read failed: Die 
Verbindung wurde vom Kommunikationspartner zurückgesetzt)
   0x0000:  43 4f 4e 4e 45 43 54 1e 1e 41 6e 64 72 6f 69 64 CONNECT..Android
   0x0010:  1e 41 6e 64 72 6f 69 64 1e 32 2e 30 2e 36 1e 30 .Android.2.0.6.0
   0x0020:  04 .
mmserver[25175]: [192.168.254.67] session ended

Kind regards

Walter

Am 18.02.2014 18:12, schrieb Jim:

**Comment by anoved** [comment]
I am going to see if Mobile Mouse will run on the Android developer emulator, or instead perhaps borrow an Android device from a friend. I'd like to fix these bugs but will understand the problem better if I can reproduce them myself. Thanks for your patience.

**Comment by anoved** [comment]
One more thing to try if any of you are willing and able: can you test whether the current Android app exhibits the same connection problems with the official Mac or PC server program? Some of the recent reviews on the Google Play page seem to describe similar problems establishing a connection. Just trying to rule out the possibility that the app itself is buggy. https://play.google.com/store/apps/details?id=com.mobilemouse.free

**Comment by anoved** [comment]
Well, I got a chance to borrow an Android tablet and to confirm this issue myself.

The good news: by trial and error, I identified what was causing the connection problem. mmserver supports customizable hotkey actions, and sends the hotkey names to the client upon connection so that the names can be displayed in the app. This functionality does not seem to be supported by the Android app (or at least not the Lite version), so the unexpected message causes the connection to fail. mmserver now refrains from sending the hotkey names if the client is an Android app. As a result, the Android app can connect!

The bad news: there appear to be other discrepancies between the recent Android and iOS versions. For example, it looks like the `KEY` and `SCROLL` packet formats are divergent. At least basic mouse movement and clicking works.
